test/emmy/tape_test.cljc

```diff
@@ -0,0 +1,495 @@
+      (is (= [z b a y x]
+             (t/topological-sort z))
+          "topological-sort returns the graph in reverse order."))))
+
+(deftest lifted-fn-tests
+  (letfn [(breaks? [f x]
+            (is (thrown? #?(:clj Exception :cljs js/Error)
+                         ((t/gradient f) x))))]
+
+    (testing "function inputs throw, as they can't be perturbed"
+      (breaks? g/square g/sin))
+
+    (testing "function inputs throw, as they can't be perturbed"
+      (breaks? (fn [{:keys [x]}] (g/square x))
+               {:x 1}))
+
+    (checking "integer-discontinuous derivatives work" 100 [x sg/real]
+              (if (v/integral? x)
+                (do (breaks? g/floor x)
+                    (breaks? g/ceiling x)
+                    (breaks? g/integer-part x)
+                    (breaks? g/fractional-part x))
+
+                (do (is (zero? ((t/gradient g/floor) x)))
+                    (is (zero? ((t/gradient g/ceiling) x)))
+                    (is (zero? ((t/gradient g/integer-part) x)))
+                    (is (g/one? ((t/gradient g/fractional-part) x)))))))
+
+  (testing "lift-n"
+    (let [*   (t/lift-n g/* (fn [_] 1) (fn [_ y] y) (fn [x _] x))
+          Df7 (t/gradient
+               (fn x**7 [x] (* x x x x x x x)))
+          Df1 (t/gradient *)
+          Df0 (t/gradient (fn [_] (*)))]
+      (is (v/= '(* 7 (expt x 6))
+               (g/simplify (Df7 'x)))
+          "functions created with lift-n can take many args (they reduce via the
+          binary case!)")
+
+      (is (= 1 (Df1 'x))
+          "single-arity acts as id.")
+
+      (is (zero? (Df0 'x))
+          "zero-arity acts as constant")))
+
+  (testing "exercise some of the lifted fns by comparing them to numeric
+  derivatives."
+    (let [f (fn [x]
+              (g/+ (g/* (g/cos x) (g/sin x))
+                   (g/+ (g/sin x) (g/expt x 2))
+                   (g/+ (g/sin x) x)
+                   (g/log (g/abs x))))
+          Df         (t/gradient f)
+          Df-numeric (D-numeric f)]
+      (with-comparator (v/within 1e-6)
+        (checking "exercise some lifted fns" 100
+                  [n (gen/double*
+                      {:infinite? false
+                       :NaN? false
+                       :min 1
+                       :max 100})]
+                  (is (ish? (Df-numeric n)
+                            (Df n))
+                      "Does numeric match autodiff?"))))))
+
+(deftest amazing-bug
+  (testing "alexey's amazing bug"
+    (let [shift (fn [offset]
+                  (fn [g]
+                    (fn [a]
+                      (g (g/+ a offset)))))
+          f-hat ((t/gradient shift) 3)]
+      (is (= (g/exp 8) ((f-hat g/exp) 5))
+          "Nothing tough expected in this case.")
+
+      ;; We'll update this once we fix the amazing bug for reverse-mode.
+      (is (= 0 ((f-hat (f-hat g/exp)) 5))
+          "This is the amazing bug, and SHOULD actually equal (g/exp 11)."))))
+
+(deftest sinc-etc-tests
```

Comment:
An idea (for the future!) might be to probabilistically generate smallish expression trees out of safely-differentiable components (+, -, *, exp) and compare forward, backward, numerical results